parlai/agents/bert_ranker/bi_encoder_ranker.py

```diff
@@ -61,6 +63,56 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
                                             self.opt["type_optimization"],
                                             self.opt["learningrate"])
 
+    def set_vocab_candidates(self, shared):
+        """Load the tokens from the vocab as candidates
+
+        self.vocab_candidates will contain a [num_cands] list of strings
+        self.vocab_candidate_vecs will contain a [num_cands, 1] LongTensor
+        """
+        if shared:
+            self.vocab_candidates = shared['vocab_candidates']
+            self.vocab_candidate_vecs = shared['vocab_candidate_vecs']
+        else:
+            if 'vocab' in (self.opt['candidates'], self.opt['eval_candidates']):
+                cands = []
+                vecs = []
+                for ind in range(1, len(self.dict)):
+                    txt = self.dict[ind]
+                    cands.append(txt)
+                    vecs.append(
+                        self._vectorize_text(txt, add_start=True, add_end=True,
+                                             truncate=self.label_truncate)
+                    )
+                self.vocab_candidates = cands
+                self.vocab_candidate_vecs = padded_3d([vecs]).squeeze(0)
```

Comment:
3d pad a matrix to make it a tensor, followed by a squeeze to make it a matrix? why not normal padded_tensor here?